src/gst.rs

```diff
@@ -84,21 +97,42 @@ impl RtspServer {
         Ok(maybe_app_src)
     }
 
-    pub fn set_credentials(&mut self, user_pass: Option<(&str, &str)>) -> Result<()> {
-        let auth = user_pass.map(|(user, pass)| {
-            let auth = RTSPAuth::new();
-            /*
-            let perm = RTSPToken::new(
-                ...
-            );
-            auth.add_basic(RTSPAuth::make_basic(user, pass).as_str(), &perm);
-            */
-            // TODO TLS https://thiblahute.github.io/GStreamer-doc/gst-rtsp-server-1.0/rtsp-server.html?gi-language=c
-            auth
-        });
+    pub fn set_credentials(&self, user_pass: Option<(&str, &str)>) -> Result<()> {
+        let auth = self.server.get_auth().unwrap_or_else(|| RTSPAuth::new());
+
+        match user_pass {
+            Some((user, pass)) => {
+                    debug!("Setting credentials for user {}", user);
+                    debug!("Password is {}", pass);
+                    let token = RTSPToken::new(&[(*RTSP_TOKEN_MEDIA_FACTORY_ROLE, &"watcher")]);
+                    let basic = RTSPAuth::make_basic(user, pass);
+                    auth.set_supported_methods(RTSPAuthMethod::Basic);
+                    auth.add_basic(basic.as_str(), &token);
+
+                    // Now we have basic auth we stop others from watching
+                    let mut token = RTSPToken::new_empty();
+                    auth.set_default_token(Some(&mut token));
+            },
+            None => {
+                let mut token = RTSPToken::new(&[(*RTSP_TOKEN_MEDIA_FACTORY_ROLE, &"watcher")]);
+                auth.set_default_token(Some(&mut token)); //By default anyone can watch if we don't turn on basic
+            }
+        }
+
+        self.server.set_auth(Some(&auth));
+        Ok(())
+    }
+
+    pub fn set_tls(&self, cert_file: &str, client_auth: TlsAuthenticationMode) -> Result<()> {
+        debug!("Setting up TLS using {}", cert_file);
+        let auth = self.server.get_auth().unwrap_or_else(|| RTSPAuth::new());
 
-        self.server.set_auth(auth.as_ref());
+        let cert = TlsCertificate::new_from_file(&cert_file).expect("Not a valid TLS certificate or not found.");
+        auth.set_tls_certificate(Some(&cert));
+        auth.set_tls_authentication_mode(client_auth);
```

Comment:
Does this line prohibit TLS _and_ basic auth? I would think that's a valid combination.

Reply:
By default when you turn on TLS you get auth turned on that stops you access anything unauthorized.

- We first set up TLS which stop everything
- When setting up the factory we add permissions so that anyone with the watcher token to play
- We then setup users
   - if there is a username we permit them only by giving only them the `watcher` token
   - If there is no username we permit anyone we give the default user the `watcher` token so that all can watch

I have set it up so that  you are able to choose TLS or Basic or TLS+Basic.

I may have missunderstood your question... Where you actually referring to `auth.set_tls_authentication_mode(client_auth);`

Reply:
```
auth.set_tls_authentication_mode(client_auth);
``
As I understand this... possible wrong:
- `client_auth = required`
    - Means I want the client to have a PEM certificate too. That cert must be signed by me. No cert from me, no access.
    - This is on top of basic auth, and adds extra security. I like to think of it as the equlivalent of an `sshkey` being required by the server.
- `client_auth = requested`
    - Means I accept PEMs signed by me or someone else, you don't need one but it would be nice.
- `client_auth = none`
    - I don't need to PEM at all and won't bother checking for one.

I am not sure of all the details about `requested` vs `none`, the docs are about sparse.

Reply:
I think you're correct about the `client_auth` semantics. I was referring to the line below, apologies. You run
```rust
auth.set_supported_methods(RTSPAuthMethod::None);
```
which, I would think that auth method is independent of whether the connection is encrypted. For example you could have an encrypted unauthenticated connection (like most HTTPS websites) or an unencrypted authenticated connection (which would be ill-advised because the passwords could be sniffed).

Reply:
In other words, let the auth-setup set the auth up, and let the encryption-setup set the encryption up.

Reply:
> Ah lack of connection may have something to do with your changes to the mount points. I specify `stream = "subStream"` and I was expecting it to mount on `name` not `name/subStream` its a bit different to what I implemented.

Didn't intentionally change the semantics, but I also like having each stream at a stable path. Sorry for the confusion.
 
> p.s. Does substream work on your camera, I thought it needed to be H264? There's no way to specify separate formats when using "both". It used to force it to always be H264 on the `subStream` but this is gone now. Will you be adding auto detection :)?

Yes it did and yes I will, soon (tm). It doesn't surprise me that there's a single codec per camera; usually the SoCs they use have one powerful encoder and that's all.

Reply:
Yes that was the general idea setting up basic should be done in the Auth set up. Originally I was hoping that setting it to none would restore the default functionality of unauthenticated can do anything but that was not the case. Having tls and no auth without  setting up default token and presmissions in factory, means we are encrypted but no-one ace access it haha. You have to always run the user set up and factory permissions with tls.

Reply:
I think I should remove it from the TLS part as it's not really helping.

Reply:
Ok so I removed the setting of basic to None during the TLS. Everything works the same even without it.